x/payment/spec/docs.md

```diff
@@ -0,0 +1,138 @@
+
+The malleated transaction that is created from metadata contained in the original `MsgWirePayForMessage`. It also burns some of the sender’s funds.
+
+## PreProcessTxs
+The malleation process occurs during the PreProcessTxs step.
+```go
+// ProcessWirePayForMessage will perform the processing required by PreProcessTxs.
+// It parses the MsgWirePayForMessage to produce the components needed to create a
+// single  MsgPayForMessage
+func ProcessWirePayForMessage(msg *MsgWirePayForMessage, squareSize uint64) (*tmproto.Message, *MsgPayForMessage, []byte, error) {
+	// make sure that a ShareCommitAndSignature of the correct size is
+	// included in the message
+	var shareCommit *ShareCommitAndSignature
+	for _, commit := range msg.MessageShareCommitment {
+		if commit.K == squareSize {
+			shareCommit = &commit
+		}
+	}
+	if shareCommit == nil {
+		return nil,
+			nil,
+			nil,
+			fmt.Errorf("message does not commit to current square size: %d", squareSize)
+	}
+
+	// add the message to the list of core message to be returned to ll-core
+	coreMsg := tmproto.Message{
+		NamespaceId: msg.GetMessageNameSpaceId(),
+		Data:        msg.GetMessage(),
+	}
+
+	// wrap the signed transaction data
+	pfm, err := msg.unsignedPayForMessage(squareSize)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return &coreMsg, pfm, shareCommit.Signature, nil
+}
+
+// PreprocessTxs fulfills the celestia-core version of the ABCI interface, by
+// performing basic validation for the incoming txs, and by cleanly separating
+// share messages from transactions
+func (app *App) PreprocessTxs(txs abci.RequestPreprocessTxs) abci.ResponsePreprocessTxs {
+	squareSize := app.SquareSize()
+	var shareMsgs []*core.Message
+	var processedTxs [][]byte
+	for _, rawTx := range txs.Txs {
+        // boiler plate
+		...
+		// parse wire message and create a single message
+		coreMsg, unsignedPFM, sig, err := types.ProcessWirePayForMessage(wireMsg, app.SquareSize())
+		if err != nil {
+			continue
+		}
+
+		// create the signed PayForMessage using the fees, gas limit, and sequence from
+		// the original transaction, along with the appropriate signature.
+		signedTx, err := types.BuildPayForMessageTxFromWireTx(authTx, app.txConfig.NewTxBuilder(), sig, unsignedPFM)
+		if err != nil {
+			app.Logger().Error("failure to create signed PayForMessage", err)
+			continue
+		}
+    ...
+	// boiler plate
+}
+```
+
+## Events
+TODO after events are added.
+
+## Parameters
+There are no parameters yet, but we might add
+- BaseFee
+- SquareSize
+- ShareSize
+
+### Usage 
+`celestia-app tx payment payForMessage <hex encoded namespace> <hex encoded data> [flags]`
+
+### Programmatic Usage
+There are tools to programmatically create, sign, and broadcast `MsgWirePayForMessages`
+```go
+// use a keyring to sign messages programmatically 
+keyringSigner, err := NewKeyringSigner(keyring, "keyring account name", "chain-id-1")
```

Comment:
Im guessing this is the keyringsinger of this module and not the general sdk. Would be good to mention as this is the most confusing part. 

May be good to open a issue on how to delete this code too.

Reply:
good point, added some more details that hopefully make this clearer. Also created an issue to look into different ways to sign malleated transactions. #192 https://github.com/celestiaorg/celestia-app/pull/187/commits/4f740a4f196dcb50aca0467c059ae943214ed1a6